# Chapter 5 - Bit Manipulation

## Most significant bit (MSB)

## Bit to int conversions

### Most sig bit

In [18]:
def msb(num, neg=True):
    sum = 0
    for i in range(0, 65):
        if sum >= num: 
            if neg: return i + 1
            return i
        sum += 2**i

num = 13
print(msb(num))


5


### Integer to binary

In [14]:
def int2bin(val, bits = None, curr='', pos=0, sep=''):
    if bits == None: bits = msb(val)
    if pos == bits: return curr

    exp = bits-pos-1
    num = 2**exp

    if val < 0:
        val = 2**exp + val
        curr += '1'+sep
    elif pos == 0 and val >= 0:
        val = val
        curr += '0'+sep
    elif val >= num:
        val = val - num
        curr += '1'
    else:
        curr += '0'
    result = int2bin(val, bits, curr, pos+1)

    #for i in range(0, bits - sig):
        #result = result[0:i] + '0' + result[i+1:len(result)]
    return result

n, b = -7, 4
print(n, int2bin(n,b))
n, b = -1, 4
print(n, int2bin(n,b))
n, b = 0, 4
print(n, int2bin(n,b))
n, b = 5, 4
print(n, int2bin(n,b))
n, b = 19, 10
print(n, int2bin(n,b))
n, b = -125, 12
print(n, int2bin(n,b))

-7 1001
-1 1111
0 0000
5 0101
19 0000010011
-125 111110000011


### Binary to integer

In [15]:
def bin2int(binary, hassign = True):
    bits = len(binary)
    result = 0
    maxbit = 2**(bits-1) 
    if hassign and binary[0] == '1': 
        val = bin2int(binary[1:], False)
        result = val - maxbit
        return result
    elif hassign:
        binary = binary[1:]

    for i in range(len(binary)-1, -1, -1):
        j = len(binary) - 1 - i
        b = int(binary[i])
        result += b * 2**j

    return result

b = '0101'
print(b, bin2int(b))
b = '1111'
print(b, bin2int(b))
b = '1101'
print(b, bin2int(b))
b = '11011010'
res = bin2int(b)
print(b, bin2int(b))
print(res, int2bin(res,len(b)))
b = '00010011'
print(b, bin2int(b))
b = '111110000011'
print(b, bin2int(b))

0101 5
1111 -1
1101 -3
11011010 -38
-38 11011010
00010011 19
111110000011 -125


### Integer to base number

In [45]:
def int2bin2(num, bits=None, hassign=True, base=2):
    rems = []
    curr = num
    neg = None
    if hassign and num < 0: neg = '1'
    if hassign and num >= 0: neg = '0'
    if num < 0: curr *= -1
    while curr > 0:
        rem = curr % base
        rems.append(str(rem))
        curr = curr // base
    binary = ''
    for i in range(len(rems)-1, -1, -1):
        binary += rems[i]
    pre = ''
    if neg != None: pre = neg
    if bits != None:
        bitsleft = bits - len(binary) - len(pre)
        for i in range(0, bitsleft):
            pre += '0'
    return pre+binary

num, bits = 4, 8
print(int2bin2(num, bits))
num, bits = -4, 8
print(int2bin2(num, bits))


100
00000 100
00000100
100
10000 100
10000100


In [ ]:
# https://www.youtube.com/watch?v=z9zKNiHPHGc

def float2bin(num, bits=None):


## Common bit tasks

In [16]:
# getbit: bit at i from right == 1 (boolean)
def getbit(num, i, verbose = False, bits=None):

    if verbose: 
        if bits == None: bits = msb(num)
        print('getbit')
        print('\tnum=\t', int2bin(num, bits), num)
        print('\t1 << i=\t', int2bin(1 << i, bits), 1 <<i)
        print('num & (1 << i)=\t', int2bin(num & (1 << i),bits),num & (1 << i))
        print()
    
    return ((num & (1 << i)) != 0)

b = '0101'
num = bin2int(b)
i = 0
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4, False)))
i = 1
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4)))
i = 2
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4)))

getbit
	num=	  5
	1 << i=	  1
num & (1 << i)=	  1

bin=0101 num=5 i=0 getbit=True
getbit
	num=	 0101 5
	1 << i=	 0010 2
num & (1 << i)=	 0000 0

bin=0101 num=5 i=1 getbit=False
getbit
	num=	 0101 5
	1 << i=	 0100 4
num & (1 << i)=	 0100 4

bin=0101 num=5 i=2 getbit=True


In [17]:
# setbit: set bit at i from right == 1 (new int)
def setbit(num, i, verbose=False,  bits=8,):
    if verbose: 
        if bits == None: bits = msb(num)
        print('setbit')
        print('\tnum=\t', int2bin(num, bits), num)
        print('\t1 << i=\t', int2bin(1 << i, bits), 1 <<i)
        print('num | (1 << i)=\t', int2bin(num | (1 << i),bits),num | (1 << i))
        print()
    return num | (1 << i)

b = '0101'
num = bin2int(b)
i = 1
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(setbit(num,i, 4, True)))

setbit
	num=	 0 5
	1 << i=	 0 2
num | (1 << i)=	 0 7

bin=0101 num=5 i=1 getbit=7


In [18]:
# clear bit: set bit at i from right == 0 (new int)
def clearbit(num, i):
    mask = ~(1 << i)
    return num & mask

def clearmsbthrui(num, i, verbose=False, bits = None):
    if bits == None: bits == msb(num)
    if verbose: print('num=',int2bin(num, bits))
    mask = (1 << i) - 1
    if verbose: print('mask=',int2bin(mask, bits))
    result = num & mask
    if verbose: print('result=',int2bin(result, bits),'\n')
    return result

clearmsbthrui(5,2, True, 4)

def clearithru0(num, i, verbose=False, bits=None):
    if bits == None: bits = msb(num)
    if verbose: print('num=', int2bin(num,bits))
    mask = (-1 << (i+1))
    if verbose: print('mask=',int2bin(mask,bits))
    result = num & mask
    if verbose: print('result=',int2bin(result,bits),'\n')
    return result

clearithru0(num, 1, True, 4)

num= 0101
mask= 0011
result= 0001 

num= 0101
mask= 1100
result= 0100 



4

In [19]:
def updatebit(num, int, bitval=1, verbose=True, bits=None):
    if bits == None: bits = msb(num)
    if verbose: print('num=', int2bin(num,bits))
    mask = ~(1 << i)
    if verbose: print('mask=',int2bin(mask,bits))
    nummask = num & mask
    if verbose: print('nummask=',int2bin(nummask,bits))
    vallsi = bitval << i
    if verbose: print('vallsi=',int2bin(vallsi,bits))
    result = nummask | vallsi
    if verbose: print('result=',int2bin(result,bits),'\n')

    return result

updatebit(5,1,1,True)

num= 0101
mask= 1101
nummask= 0101
vallsi= 0010
result= 0111 



7

## Questions

### 5.1 Insertion: 

You are given two 32-bit numbers, Nand M, and two bit positions, i and
j. Write a method to insert Minto N such that M starts at bit j and ends at bit i. You
can assume that the bits j through i have enough space to fit all of M. That is, if
M = 10011, you can assume that there are at least 5 bits between j and i. You would not, for
example, have j = 3 and i = 2, because M could not fully fit between bit 3 and bit 2.


EXAMPLE

Input: N = 10000000000, M 10011, i 2, j 6

Output: N 10001001100

In [21]:
def insertion(n, m, i, j, bits = None):
    if bits == None: bits = msb(n)
    print('n=', int2bin(n, bits))
    print('m=', int2bin(m, bits))
    allones = ~0
    print('allones=', int2bin(allones, bits))
    left = allones << (j+1)
    print('left=', int2bin(left, bits))
    right = (1 << i) -1 
    print('right=', int2bin(right, bits))
    mask = left | right
    print('mask=', int2bin(mask, bits), bin(mask))
    nclear = n & mask
    print('nclear=', int2bin(nclear, bits))
    mshift = m << i
    print('mshift=', int2bin(mshift, bits))
    result = nclear | mshift
    print('result=', int2bin(result, bits))



nbin = '010000011001'
mbin = '010011'
n = bin2int(nbin)
m = bin2int(mbin)
bits = len(nbin)
i = 2
j = 6

insertion(n, m, i, j, bits)

n= 010000011001
m= 000000010011
allones= 111111111111
left= 111110000000
right= 000000000011
mask= 111110000011 -0b1111101
nclear= 010000000001
mshift= 000001001100
result= 010001001101


### 5.2 Binary to String: 

Given a real number between 8 and 1 (e.g., 0.72) that is passed in as a double,
print the binary representation. If the number cannot be represented accurately in binary with at
most 32 characters, print "ERROR:'

https://www.youtube.com/watch?v=3FA8yQ0cZyE

In [10]:
def dec2bin(n):
    binary = "."
    while n > 0:
        n2 = n*2
        if n2 >= 1:
            binary += '1'
            n = n2 - 1
        else:
            binary += '0'
            n = n2
    return binary

n = 0.703125
print(dec2bin(n))
        

.101101


### 5.4 Flip Bit to Win: 

You have an integer and you can flip exactly one bit from a 0 to a 1. Write code to
find the length of the longest sequence of 1 s you could create.

EXAMPLE

Input:
1775 (or: 11011101111)

Output:
8


In [6]:
def altseqs(n):
    seqs = []
    search = 0
    count = 0
    for i in range(0, 4*8):
        if n & 1 != search:
            seqs.append(count)
            search = n & 1
            count = 0
        count += 1
        n = n >> 1
    return seqs

def consec(n):
    seqs = altseqs(n)
    maxseq = 0
    zero = 0
    prev = 0
    after = 0
    for i in range(2, len(seqs), 2):
        zero = seqs[i]
        if zero != 1: continue
        prev = seqs[i-1]
        after = 0 if i+1 > len(seqs)-1 else seqs[i+1]
        seq = prev + 1 + after
        maxseq = max(maxseq, seq)
    return maxseq
        


n = 1775
print(bin(n))
print(altseqs(n))
print(consec(n))

0b11011101111
[0, 4, 1, 3, 1, 2]
8


### 5.4 Next Number: 

Given a positive integer, print the next smallest and the next largest number that
have the same number of 1 bits in their binary representation.

### 5.6 Conversion: 

Write a function to determine the number of bits you would need to flip to convert
integer A to integer B.


EXAMPLE

Input: 29 (or: 111101), 15 (or: 101111)

Output: 2

In [19]:
def conv(a, b):
    count = abs(msb(a) - msb(b))
    c = a & b
    while c > 0:
        if c & 1 == 0: count += 1
        c = c >> 1
    return count

def msb(num):
    sum = 0
    for i in range(0, 65):
        if sum >= num: 
            return i
        sum += 2**i

a, b = 13, 10
print(conv(a,b))

a, b = 29, 15
print(conv(a,b))

3
2


### 5.7 Pairwise Swap: 

Write a program to swap odd and even bits in an integer with as few instructions as
possible (e.g., bit 13 and bit 1 are swapped, bit 2 and bit 3 are swapped, and so on).

In [42]:
def swap(n):
    result = 0
    for i in range(0, 32, 2):
        a = n >> i
        if a == 0: return result
        b = a & 3
        c = b << 1 & 3 | (b >> 1 & 1)
        d = c << i
        print('a',bin(a),'b', bin(b), 'c', bin(c), 'd', bin(d))
        result = result | d
    return result



n = 13
print(bin(n), bin(swap(n)))

a 0b1101 b 0b1 c 0b10 d 0b10
a 0b11 b 0b11 c 0b11 d 0b1100
0b1101 0b1110


In [49]:
def swap2(n):
    leftshift = n*(2**1)
    rightshift = n//(2**1)
    print(bin(leftshift), bin(rightshift))
    return leftshift | rightshift

n = 13
print(bin(n), bin(swap2(n)))

0b11010 0b110
0b1101 0b11110
